### PR TITLE
p2p: Implement P2P Bucket data structure

### PR DESCRIPTION
**Pull request by SyntheticBird45**
### What

See: https://github.com/Cuprate/cuprate/issues/178

This PR implements a "Bucket" data structure that is a collection of data that discriminates its items into "buckets" (vector of size N) following a defined function.

- Implements Bucket data structure and Bucketable trait
- Implements Bucketable for Ipv4Addr
- Added the crate to the workspace dependencies
- Added `arrayvec` as a dependency

### Why

The end goal is to use this structure to store diverse IPs of different /16 subnets.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -51,35 +52,37 @@ opt-level = 3
...
-cuprate-database-service    = { path = "storage/service"        ,default-features = false}
-cuprate-txpool              = { path = "storage/txpool"         ,default-features = false}
-cuprate-pruning             = { path = "pruning"                ,default-features = false}
-cuprate-test-utils          = { path = "test-utils"             ,default-features = false}
-cuprate-types               = { path = "types"                  ,default-features = false}
-cuprate-json-rpc            = { path = "rpc/json-rpc"           ,default-features = false}
-cuprate-rpc-types           = { path = "rpc/types"              ,default-features = false}
-cuprate-rpc-interface       = { path = "rpc/interface"          ,default-features = false}
+cuprate-fast-sync           = { path = "consensus/fast-sync"  ,default-features = false}
+cuprate-consensus-rules     = { path = "consensus/rules"      ,default-features = false}
```
Comment (Boog900) [inline Cargo.toml:56]: Can you revert the changes to the unrelated lines please

File: p2p/bucket/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "cuprate-p2p-bucket"
+version = "0.1.0"
+edition = "2021"
+
```
Comment (Boog900) [inline p2p/bucket/Cargo.toml:5]: This will need license etc, can you copy from another crate and edit appropriately

File: p2p/bucket/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+/// A collection data structure discriminating its unique items
+/// with a specified method. Enable a fair distribution of the
+/// items based on their discriminants when popped.
+pub struct Bucket<const N: usize, I: Bucketable> {
+    /// The storage of the bucket
+    storage: BTreeMap<I::Discriminant, ArrayVec<I, N>>,
+    /// Internal round-robin state
+    round_robin: usize,
```
Comment (Boog900) [inline p2p/bucket/src/lib.rs:32]: for now can you remove this round_robin pop impl please.

It should be done as a separate struct which impls `Iterator`, however until we need it I would just remove this.

**Review by Boog900 — CHANGES_REQUESTED** [review]

File: p2p/bucket/src/lib.rs (hunk 3 above)
Comment (SyntheticBird45) [reply p2p/bucket/src/lib.rs:32]: I'm actually planning on passing to the Address book PR after that. Shouldn't it be preferable to make that iterator structure in this PR instead ?

Comment (Boog900) [reply p2p/bucket/src/lib.rs:32]: The bucket structure is more for making sure we store a wide range of IPs, not that we select a wide range of IPs, having this round_robin makes it more likely for consecutive subnets to be all the peers we connect to.

Just randomly selecting peers across the whole structure would be better IMO, but this can be considered later.

File: p2p/bucket/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    pub fn push(&mut self, item: I) {
+        let discriminant = item.discriminant();
+
+        if let Some(vec) = self.storage.get_mut(&discriminant) {
+            // Check if the element already exists
+            let already_exist = vec.iter().any(|v| &item == v);
+            if !already_exist {
+                // Push the new element
+                vec.push(item);
+            }
```
Comment (hinto-janai) [inline p2p/bucket/src/lib.rs:60]:
```suggestion
            // Push the item if it doesn't exist.
            if !vec.contains(&item) {
                vec.push(item);
            }
```

File: p2p/bucket/src/lib.rs (hunk 3 above)
Comment (SyntheticBird45) [reply p2p/bucket/src/lib.rs:32]: Fair enough.

File: p2p/bucket/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+//!
+//! A collection data structure that discriminates its unique items and place them into "buckets"
+//!
+//! The item must implement the [`Bucketable`] trait that defines how to create the discriminant
+//! from the item type. The data structure will internally contain any item into "buckets" or vectors
+//! of sized capacity N that regroup all the stored items with this specific discriminant.
+//!
+//! A practical example of this data structure is for storing N amount of IP discriminated by their subnets.
+//! You can store in each "buckets" corresponding to a /16 subnet up to N IPs of that subnet.
+//!
```
Comment (hinto-janai) [inline p2p/bucket/src/lib.rs:11]: An example usage test-doc would be nice.

File: p2p/bucket/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    /// Will attempt to remove an item from the bucket
+    pub fn remove(&mut self, item: &I) -> Option<I> {
+        self.storage.get_mut(&item.discriminant()).and_then(|vec| {
+            let find = vec
+                .iter()
+                .enumerate()
+                .filter(|(_, v)| &item == v)
+                .map(|(i, _)| i)
+                .last();
+            find.map(|index| vec.swap_remove(index))
```
Comment (hinto-janai) [inline p2p/bucket/src/lib.rs:78]:
```suggestion
            vec.iter()
                .enumerate()
                .filter_map(|(i, v)| (item == v).then_some(i))
                .last()
                .map(|index| vec.swap_remove(index))
```

File: p2p/bucket/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+impl<const N: usize, I: Bucketable> Default for Bucket<N, I> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Bucketable for Ipv4Addr {
+    // We are discriminating by /16 subnets
```
Comment (hinto-janai) [inline p2p/bucket/src/lib.rs:111]:
```suggestion
    /// We are discriminating by `/16` subnets.
```

**Review by hinto-janai — COMMENTED** [review]
Can you add this crate to here?: https://github.com/Cuprate/cuprate/blob/main/books/architecture/src/appendix/crates.md

File: p2p/bucket/src/lib.rs (hunk 3 above)
Comment (SyntheticBird45) [reply p2p/bucket/src/lib.rs:32]: If I remove the current pop there is no method to actually extract an IP address from the storage. Do you want me to make the `storage` field public ? or adapt the current `pop` to take a random item (i know this contradict your statement of needing an iterator but I'm going to implement anyway in the next PR since well I'll need it)?

Comment (Boog900) [reply p2p/bucket/src/lib.rs:32]: Yeah adding a `pop` method that selects  a random peer would be good - although it should leave the peer in structure and just return a reference

File: p2p/bucket/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+//! assert_eq!(2, bucket.len_bucket(&[96_u8,96_u8]).unwrap());
+//!
+//! // Push a new IP from another subnet
+//! bucket.push("127.0.0.1".parse().unwrap());
+//! assert_eq!(3, bucket.len());
+//! assert_eq!(2, bucket.len_bucket(&[96_u8,96_u8]).unwrap());
+//! assert_eq!(1, bucket.len_bucket(&[127_u8,0_u8]).unwrap());
+//!
+//! // Attempting to push a new IP within `96.96.0.0/16` bucket will panic.
+//! ```
```
[the remark this answers is not in the document]
Comment (SyntheticBird45) [reply p2p/bucket/src/lib.rs:34]: Yes, It will. 
Yes, It is a bad API, because it imply you have to use `len_bucket()` before use.
Yes, I would like to replace it with `try_push` or `push` but return a result or option.

[the remark this answers is not in the document]
Comment (Boog900) [reply p2p/bucket/src/lib.rs:34]: Ah it's an array vec, yes it should return an option, it should not panic

File: p2p/bucket/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    /// Will attempt to remove an item from the bucket.
+    pub fn remove(&mut self, item: &I) -> Option<I> {
+        self.storage.get_mut(&item.discriminant()).and_then(|vec| {
+            vec.iter()
+                .enumerate()
+                .filter_map(|(i, v)| (item == v).then_some(i))
+                .last()
+                .map(|index| vec.swap_remove(index))
```
Comment (Boog900) [inline p2p/bucket/src/lib.rs:122]: I think this could probably be made more efficient, maybe by using `find_map` instead of `filter_map`

**Review by Boog900 — APPROVED** [review]
